saleor/plugins/avatax/__init__.py

```diff
@@ -242,24 +249,71 @@ def get_checkout_lines_data(
             discounts,
-        ).gross.amount
+        )
+
+        if tax_included:
+            undiscounted_amount = prices_data.undiscounted_price.gross.amount
+            price_amount = prices_data.price.gross.amount
+            price_with_voucher_amount = prices_data.price_with_voucher.gross.amount
+        else:
+            undiscounted_amount = prices_data.undiscounted_price.net.amount
+            price_amount = prices_data.price.net.amount
+            price_with_voucher_amount = prices_data.price_with_voucher.net.amount
+
         append_line_to_data(
             data=data,
             quantity=line_info.line.quantity,
-            amount=amount,
+            amount=undiscounted_amount,
             # This is a workaround for Avatax and sending a lines with amount 0. Like
             # order lines which are fully discounted for some reason. If we use a
             # standard tax_code, Avatax will raise an exception: "When shipping
             # cross-border into CIF countries, Tax Included is not supported with mixed
             # positive and negative line amounts."
-            tax_code=tax_code if amount else DEFAULT_TAX_CODE,
+            tax_code=tax_code if undiscounted_amount else DEFAULT_TAX_CODE,
             item_code=line_info.variant.sku,
             name=name,
+            tax_included=tax_included,
         )
+        if undiscounted_amount != price_amount:
+            append_line_to_data(
+                data=data,
+                quantity=line_info.line.quantity,
+                amount=price_amount,
+                # This is a workaround for Avatax and sending a lines with amount 0.
+                # Like order lines which are fully discounted for some reason. If we use
+                # a standard tax_code, Avatax will raise an exception: "When shipping
+                # cross-border into CIF countries, Tax Included is not supported with
+                # mixed positive and negative line amounts."
+                tax_code=tax_code if undiscounted_amount else DEFAULT_TAX_CODE,
+                item_code=line_info.variant.sku,
+                name=name,
+                tax_included=tax_included,
+                # ref1 is custom field which we use to determine for which line we
+                # applied sale_discount
+                ref1=line_info.variant.sku,
+            )
+        if price_amount != price_with_voucher_amount:
+            append_line_to_data(
+                data=data,
+                quantity=line_info.line.quantity,
+                amount=price_with_voucher_amount,
+                # This is a workaround for Avatax and sending a lines with amount 0.
+                # Like order lines which are fully discounted for some reason. If we use
+                # a standard tax_code, Avatax will raise an exception: "When shipping
+                # cross-border into CIF countries, Tax Included is not supported with
+                # mixed positive and negative line amounts."
```

Comment:
Do we need the commend to be duplicated? Also since the args are very similar maybe we could create a dict with args and in the "if" condition just append different refs?

Reply:
Great! Makes huge sense. Thanks!